extensions/src/main/java/io/lionweb/repoclient/RepoSerialization.java

```diff
@@ -0,0 +1,261 @@
+    }
+
+    try (FileInputStream fis = new FileInputStream(zip);
+        ZipInputStream zis = new ZipInputStream(fis, StandardCharsets.UTF_8)) {
+
+      ZipEntry entry;
+      final int BUFFER_SIZE = 32 * 1024;
+      byte[] buffer = new byte[BUFFER_SIZE];
+      while ((entry = zis.getNextEntry()) != null) {
+        if (!entry.getName().toLowerCase().endsWith(".json")) {
+          continue;
+        }
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        int bytesRead;
+        while ((bytesRead = zis.read(buffer)) != -1) {
+          baos.write(buffer, 0, bytesRead);
+        }
+
+        String content = new String(baos.toByteArray(), StandardCharsets.UTF_8);
+
+        LowLevelJsonSerialization lowLevelJsonSerialization = new LowLevelJsonSerialization();
+        SerializedChunk serializedChunk =
+            lowLevelJsonSerialization.deserializeSerializationBlock(content);
+        SerializedClassifierInstance root =
+            serializedChunk.getClassifierInstances().stream()
+                .filter(n -> n.getParentNodeID() == null)
+                .findFirst()
+                .get();
+        root.clearContainments();
+        SerializedChunk limitedSerializedChunk =
+            groupNodesIntoSerializationBlock(
+                Collections.singletonList(root), apiClient.getLionWebVersion());
+        String limitedJson =
+            lowLevelJsonSerialization.serializeToJsonString(limitedSerializedChunk);
+
+        apiClient.rawCreatePartitions(limitedJson);
+
+        apiClient.rawStore(content);
+
+        zis.closeEntry();
+      }
+    }
+  }
+
+  /**
+   * Upload all the content of a zip to a given repository, using the bulkImport operation (and not
+   * the standard bulk operations). All the zip is examined, looking for entries with extension
+   * ".json" (ignoring case).
+   */
+  public void uploadZipToRepoUsingBulkImport(AdditionalAPIClient apiClient, File zip)
+      throws IOException {
+    if (!zip.isFile()) {
+      throw new IllegalArgumentException(
+          "Provided path is not a valid zip file: " + zip.getAbsolutePath());
+    }
+
+    final int BUFFER_SIZE = 32 * 1024;
+
+    LowLevelJsonSerialization lowLevelJsonSerialization = new LowLevelJsonSerialization();
+    BulkImport bulkImport = new BulkImport();
+
+    try (FileInputStream fis = new FileInputStream(zip);
+        ZipInputStream zis = new ZipInputStream(fis, StandardCharsets.UTF_8)) {
+
+      ZipEntry entry;
+      byte[] buffer = new byte[BUFFER_SIZE];
+
+      while ((entry = zis.getNextEntry()) != null) {
+        if (!entry.getName().toLowerCase().endsWith(".json")) {
+          continue;
+        }
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        int bytesRead;
+        while ((bytesRead = zis.read(buffer)) != -1) {
+          baos.write(buffer, 0, bytesRead);
+        }
+
+        // Write to temp file (needed because deserializeSerializationBlock expects a File)
```

Comment:
This feels like `deserializeSerializationBlock` should be refactored to work on streams

Reply:
This is now https://github.com/LionWeb-io/lionweb-java/issues/218